### PR TITLE
Provider Interface

### PR DESCRIPTION
**Pull request by jackzampolin**
With the need to support additional chain types, most of the functionality in the golang relayer needs to be put behind an interface so that we can implement that interface for different chain types (i.e. substrate). This is an initial effort to define that interface.

### CODE REVIEW

File: relayer/provider/cosmos/provider.go [hunk 1]
```diff
@@ -496,10 +497,12 @@ func (cp *CosmosProvider) SendMessages(msgs []provider.RelayerMessage) (*provide
 		return nil, false, err
 	}
 
-	// TODO helper/wrapper function for TxResponse->RelayerTxResponse
+	// helper/wrapper function for TxResponse->RelayerTxResponse?
```
Comment (jackzampolin) [inline relayer/provider/cosmos/provider.go]: Yes I think we want this wrapper function

File: relayer/provider/provider.go [hunk 2]
```diff
@@ -68,7 +69,7 @@ type QueryProvider interface {
 	QueryUpgradedClient(height int64) (*clienttypes.QueryClientStateResponse, error)
 	QueryUpgradedConsState(height int64) (*clienttypes.QueryConsensusStateResponse, error)
 	QueryConsensusState(height int64) (ibcexported.ConsensusState, int64, error)
-	QueryClients() ([]*clienttypes.IdentifiedClientState, error)
+	QueryClients() ([]clienttypes.IdentifiedClientState, error)
```
Comment (jackzampolin) [inline relayer/provider/provider.go]: y no pointer?

Comment (jtieri) [reply relayer/provider/provider.go]: The returned data, `IdentifiedClientState` is exposed via a struct found in `ibc-go` which is an array of structs vs. array of pointers

https://github.com/cosmos/relayer/blob/5d02ee2710ab673dbc8e4661f03572d82d9b8389/relayer/provider/cosmos/provider.go#L734

```
// QueryClientStatesResponse is the response type for the Query/ClientStates RPC
// method.
type QueryClientStatesResponse struct {
	// list of stored ClientStates of the chain.
	ClientStates IdentifiedClientStates `protobuf:"bytes,1,rep,name=client_states,json=clientStates,proto3,castrepeated=IdentifiedClientStates" json:"client_states"`
	// pagination response
	Pagination *query.PageResponse `protobuf:"bytes,2,opt,name=pagination,proto3" json:"pagination,omitempty"`
}
```

File: relayer/provider/cosmos/provider.go (hunk 1 above)
Comment (jtieri) [reply relayer/provider/cosmos/provider.go]: I'll get that added in here!

**Comment by jackzampolin** [comment]
Just going to send it

**Pull request merged**